Resources/Public/Javascript/OpenLayers/ImageProcessingUtility.js

```diff
@@ -0,0 +1,225 @@
+    var luG = 0.6094;
+    var luB = 0.0820;
+
+    var az = (1 - sv)*luR + sv;
+    var bz = (1 - sv)*luG;
+    var cz = (1 - sv)*luB;
+    var dz = (1 - sv)*luR;
+    var ez = (1 - sv)*luG + sv;
+    var fz = (1 - sv)*luB;
+    var gz = (1 - sv)*luR;
+    var hz = (1 - sv)*luG;
+    var iz = (1 - sv)*luB + sv;
+
+    for(var i = 0; i < data.length; i += 4)
+    {
+      var red = data[i]; // Extract original red color [0 to 255]. Similarly for green and blue below
+      var green = data[i + 1];
+      var blue = data[i + 2];
+
+      var saturatedRed = (az*red + bz*green + cz*blue);
+      var saturatedGreen = (dz*red + ez*green + fz*blue);
+      var saturateddBlue = (gz*red + hz*green + iz*blue);
+
+      data[i] = saturatedRed;
+      data[i + 1] = saturatedGreen;
+      data[i + 2] = saturateddBlue;
+    }
+
+    return imgData;
+  },
+
+  'brightness': function (imgData, value) {
+    var data = imgData.data;
+    var brightnessMul = value; // brightness multiplier
+
+    for(var i = 0; i < data.length; i += 4)
+    {
+      data[i] = brightnessMul * data[i];
+      data[i + 1] = brightnessMul * data[i + 1];
+      data[i + 2] = brightnessMul * data[i + 2];
+    }
+
+    return imgData;
+  },
+
+  'invert': function (imgData) {
+    var data = imgData.data;
+    for (var i = 0; i < data.length; i += 4) {
+      data[i]     = 255 - data[i];     // red
+      data[i + 1] = 255 - data[i + 1]; // green
+      data[i + 2] = 255 - data[i + 2]; // blue
+    }
+    return imgData;
+  },
+
+  'hue': function(imgData, value) {
+    var data = imgData.data;
+
+    var hue = value / 360;
+
+    var rgb = [0, 0, 0];
+    var hsv = [0.0, 0.0, 0.0];
+    for (var i = 0; i < data.length; i += 4) {
+      // retrieve r,g,b (! ignoring alpha !)
+      var r = data[i];
+      var g = data[i + 1];
+      var b = data[i + 2];
+      // convert to hsv
+      this.rgbToHsv(r, g, b, hsv);
+      // change color if hue near enough from tgtHue
+      // adjust hue
+      // ??? or do not add the delta ???
+      hsv[0] = hue  //+ hueDelta;
+      // convert back to rgb
+      this.hsvToRgb(rgb, hsv);
+      // store
+      data[i] = rgb[0];
+      data[i + 1] = rgb[1];
+      data[i + 2] = rgb[2];
+    }
+
+    return imgData;
+  },
+
+  'rgbToHsv': function (r, g, b, hsv) {
+    var K = 0.0,
+      swap = 0;
+    if (g < b) {
+      swap = g;
+      g = b;
+      b = swap;
+      K = -1.0;
+    }
+    if (r < g) {
+      swap = r;
+      r = g;
+      g = swap;
+      K = -2.0 / 6.0 - K;
+    }
+    var chroma = r - (g < b ? g : b);
+    hsv[0] = Math.abs(K + (g - b) / (6.0 * chroma + 1e-20));
+    hsv[1] = chroma / (r + 1e-20);
+    hsv[2] = r;
+  },
+
+  'hsvToRgb': function (rgb, hsv) {
+    var h = hsv[0];
+    var s = hsv[1];
+    var v = hsv[2];
+
+    // The HUE should be at range [0, 1], convert 1.0 to 0.0 if needed.
+    if (h >= 1.0) h -= 1.0;
+
+    h *= 6.0;
+    var index = Math.floor(h);
+
+    var f = h - index;
```

Comment:
f? p? q? t? Could we use here more specific names?

Reply:
these formulas are taken from various image processing libraries. thats a lot of math magic, which i dont fully understand. but they are working. Same in the other comments.